torch/lib/THCUNN/generic/ClassNLLCriterion.cu

```diff
@@ -46,6 +43,37 @@ void THNN_(ClassNLLCriterion_updateOutput)(
             " but got weight tensor of shape: %s", n_classes, s1.str);
   }
 
+  if (!reduce && n_dims == 2) {
+    THCTensor_(resize1d)(state, output, batch_size);
+    bool not_contiguous_weights = weights && !THCTensor_(isContiguous)(state, weights);
```

Comment:
Use:

```
weights = THCTensor_(newContiguous)(state, weights)
...
THCTensor_(free)(state, weights)
```

i.e. don't guard it with a contiguous check. `newContiguous` will already avoid the copy and just incref the tensor if it's already contiguous.